synapse/lib/stormtypes.py

```diff
@@ -1354,6 +1354,123 @@ def __init__(self, path=None):
+            'add': self._methViewAdd,
+            'del': self._methViewDel,
+            'fork': self._methViewFork,
+            'get': self._methViewGet,
+            'list': self._methViewList,
+            'merge': self._methViewMerge,
+        })
+
+    async def _methViewAdd(self, layers):
+        '''
+        Add a view to the cortex.
+        '''
+        self.runt.reqAllowed(('storm', 'view', 'add'))
+
+        iden = s_common.guid()
+        view = await self.runt.snap.addView(iden, layers)
+
+        if view is None:
+            mesg = f'Failed to add view.'
+            raise s_exc.StormRuntimeError(mesg=mesg, iden=iden, layers=layers)
+
+        return View(view, path=self.path)
+
+    async def _methViewDel(self, iden):
+        '''
+        Delete a view in the cortex.
+        '''
+        view = self.runt.snap.getView(iden)
+        if view is None:
+            mesg = f'No view with iden: {iden}'
+            raise s_exc.NoSuchIden(mesg=mesg)
+
+        if view is self.runt.snap.core.view:
+            mesg = f'Deleting the main view is not permitted.'
+            raise s_exc.StormRuntimeError(mesg=mesg, iden=iden)
+
+        if not view.info.get('owner') == self.runt.user.iden:
+            self.runt.reqAllowed(('storm', 'view', 'del'))
+
+        await self.runt.snap.delView(iden=view.iden)
+
+        return True
+
+    async def _methViewFork(self, iden):
+        '''
+        Fork a view in the cortex.
+        '''
+        self.runt.reqAllowed(('storm', 'view', 'add'))
+
+        view = self.runt.snap.getView(iden)
+        if view is None:
+            mesg = f'No view with iden: {iden}'
+            raise s_exc.NoSuchIden(mesg=mesg)
+
+        newview = await view.fork(owner=self.runt.user.iden)
+
+        return View(newview, path=self.path)
+
+    async def _methViewGet(self, iden=None):
+        '''
+        Retrieve a view from the cortex.
+        '''
+        self.runt.reqAllowed(('storm', 'view', 'get'))
+
+        view = self.runt.snap.getView(iden)
+        if view is None:
+            mesg = f'No view with iden: {iden}'
+            raise s_exc.NoSuchIden(mesg=mesg)
+
+        return View(view, path=self.path)
+
+    async def _methViewList(self):
+        '''
+        List the views in the cortex.
+        '''
+        self.runt.reqAllowed(('storm', 'view', 'get'))
+
+        views = self.runt.snap.listViews()
+
+        return [View(view, path=self.path) for view in views]
+
+    async def _methViewMerge(self, iden):
+        '''
+        Merge a forked view back into its parent.
+
+        When complete, the view is deleted.
+        '''
+        view = self.runt.snap.getView(iden)
+        if view is None:
+            mesg = f'No view with iden: {iden}'
+            raise s_exc.NoSuchIden(mesg=mesg)
+
+        if not view.info.get('owner') == self.runt.user.iden:
+            self.runt.reqAllowed(('storm', 'view', 'del'))
+
+        if await view.mergeAllowed(self.runt.user):
```

Comment:
This might take a while, so we should probably lock/readonly the forked write layer first (and then un-readonly if fail)